### ISSUE TITLE
Compiler crash with error "no-symbol does not have an owner"

### ISSUE THREAD

**Issue by scabug** [scala/bug#7891]
Attached code crashes the compiler. Tested with a couple of recent versions: 2.10.2, 2.10.3, 2.11.0-M5. It does compile with 2.9.3.

The error message from 2.10.3:
```
[error] no-symbol does not have an owner
unhandled exception while transforming CompilerErrorTest.scala
[error] uncaught exception during compilation: scala.reflect.internal.FatalError
[trace] Stack trace suppressed: run last compile:compile for the full output.
[error] (compile:compile) scala.reflect.internal.FatalError:
[error]      while compiling: E:\Users\Jarek\local\openni-samples\src\main\scala\CompilerErrorTest.scala
[error]         during phase: global=explicitouter, atPhase=erasure
[error]      library version: version 2.10.3
[error]     compiler version: version 2.10.3
[error]   reconstructed args: -bootclasspath [...] -classpath [...]
[error]
[error]   last tree to typer: This(anonymous class $anon)
[error]               symbol: anonymous class $anon (flags: final <inconstructor>)
[error]    symbol definition: final class $anon extends KeyListener
[error]                  tpe: CompilerErrorTest.CompilerErrorTest.$anon.type
[error]        symbol owners: anonymous class $anon -> constructor CompilerErrorTest -> class CompilerErrorTest -> class CompilerErrorTest -> package <empty>
[error]       context owners: value arg0 -> method keyPressed -> anonymous class $anon -> constructor CompilerErrorTest -> class CompilerErrorTest -> class CompilerErrorTest -> package <empty>
[error]
[error] == Enclosing template or block ==
[error]
[error] DefDef( // def keyPressed(arg0: java.awt.event.KeyEvent): Unit
[error]   <method>
[error]   "keyPressed"
[error]   []
[error]   // 1 parameter list
[error]   ValDef( // arg0: java.awt.event.KeyEvent
[error]     <param> <triedcooking>
[error]     "arg0"
[error]     <tpt> // tree.tpe=java.awt.event.KeyEvent
[error]     <empty>
[error]   )
[error]   <tpt> // tree.tpe=Unit
[error]   If( // tree.tpe=Unit
[error]     Apply( // def ==(x: Int): Boolean in class Int, tree.tpe=Boolean
[error]       arg0.getKeyCode()."$eq$eq" // def ==(x: Int): Boolean in class Int, tree.tpe=(x: Int)Boolean
[error]       27
[error]     )
[error]     Apply( // final def mShouldRun_=(x$1: Boolean): Unit in class CompilerErrorTest, tree.tpe=Unit
[error]       CompilerErrorTest.this."mShouldRun_$eq" // final def mShouldRun_=(x$1: Boolean): Unit in class CompilerErrorTest, tree.tpe=(x$1: Boolean)Unit
[error]       false
[error]     )
[error]     ()
[error]   )
[error] )
[error]
[error] == Expanded type of tree ==
[error]
[error] ThisType(anonymous class $anon)
[error]
[error] no-symbol does not have an owner
```

**Comment by diesalbla** [comment]
@som-snytt The compile crash seems to be solved in the `2.13.x` branch. When running `sbt scala` and introducing the code of Paul Philips above,  I get the following: 
```Scala
scala> :paste
// Entering paste mode (ctrl-D to finish)

class B { self =>
  def this(x: Any) { this() ; new Object { self } }
}

// Exiting paste mode, now interpreting.

         def this(x: Any) { this() ; new Object { self } }
                                                  ^
On line 2: warning: a pure expression does nothing in statement position; multiline expressions may require enclosing parentheses
         def this(x: Any) { this() ; new Object { self } }
                                         ^
On line 2: error: Internal error: unable to find the outer accessor symbol of <$anon: Object>
```

I would need to confirm if that is the intended behavior.

**Comment by som-snytt** [comment]
I know there have been improvements to whether to hold onto an outer pointer; but there was also a recent example where it didn't correctly capture it, maybe one of your regression tests.

The anon reference to enclosing template should either receive that reference (in which case the reference is pure and can be elided) or the outer is an object and the reference is not pure because it invokes an initializer; I don't know if that's handled correctly, but it seems to me it should be one case or the other. I suspect the messaging is incorrect, but it's also useful to submit a test that demonstrates the status quo, with a big comment that it is wrong behavior.

They used to have a "pending" category where that distinction was more obvious.
